cpmpy/tools/explain/mus.py

```diff
@@ -108,6 +111,55 @@ def do_recursion(soft, hard, delta):
     return [dmap[a] for a in core]
 
 
+def optimal_mus(soft, hard=[], weights=1, solver="ortools", hs_solver="ortools"):
+    """
+        Find an optimal MUS according to a linear objective function.
+        Works by iteratively generating correction subsets and computing optimal hitting sets to those enumerated sets.
+        For better performance of the algorithm, use an incemental solver to compute the hitting sets such as Gurobi.
+
+        Assumption-based implementation for solvers that support s.solve(assumptions=...)
+        More naive version available as `optimal_mus_naive` to use with other solvers.
+
+        CPMpy implementation loosely based on the "SMUS" algorithm from
+            Ignatiev, Alexey, et al. "Smallest MUS extraction with minimal hitting set dualization."
+            International Conference on Principles and Practice of Constraint Programming. Cham: Springer International Publishing, 2015.
+        and OCUS from:
+            Gamba, Emilio, Bart Bogaerts, and Tias Guns. "Efficiently explaining CSPs with unsatisfiable subset optimization."
+            Journal of Artificial Intelligence Research 78 (2023): 709-746.
+
+    """
+
+
+    model, soft, assump = make_assump_model(soft, hard)
+    dmap = dict(zip(assump, soft))
+
+    s = cp.SolverLookup.get(solver, model)
+    # s.solution_hint(assump, [1]*len(assump)) # causes weirdness in OR-Tools: https://github.com/google/or-tools/issues/4324
+    assert s.solve(assumptions=assump) is False
+
+    # initialize hitting set solver
+    if is_num(weights):
+        weights = np.ones(len(assump), dtype=int)
+    hs_solver = cp.SolverLookup.get(hs_solver)
+    hs_solver.minimize(cp.sum(assump * np.array(weights)))
```

Comment:
you already assert s.solve() is false, so we know of one unsat core/mss; could that be used to initialize the hitting set solver?
Otherwise the simplest mcs to initialize with is simply all softs; you will need to hit at least one of those; this might 'save' one iteration (as I think now it will call the HS with no constraints so it will set all assumps to 0, instead of at least one to 1?

Reply:
Hmm, not sure how we can use the core to init the set of MCSes? We cannot derive an MCS from a core...
Agree with saving 1 iteration tho, probably cleanest is to add a constraint `cp.sum(assump) >= 1` to the hs solver at the start.
Ah but... then what if the hard constraint themselves are UNSAT? Should we check for that? It's now done implicitely in iteration 1

Reply:
hmm. I guess with the current call, if the assumps are all false it must be in the hards?